gh-80793: Edit unittest.rst on SetUpClass calls

vrpolakatcisco opened the pull request:
Signed-off-by: Vratko Polak <vrpolak@cisco.com>

<!-- gh-issue-number: gh-80793 -->
* Issue: gh-80793
<!-- /gh-issue-number -->

MakDon commented on Misc/NEWS.d/next/Documentation/2019-04-12-12-30-08.bpo-36612.8AGXVv.rst:
```diff
@@ -0,0 +1 @@
+skip news
```
```suggestion
```

MakDon commented on Doc/library/unittest.rst:
```diff
@@ -2274,9 +2274,12 @@ These must be implemented as class methods::
-:class:`TestCase` are empty.
+The implementations in :class:`TestCase` are empty.
+
+If you want the ``setUpClass`` and ``tearDownClass`` on a superclass
```
```suggestion
If you want the ``setUpClass`` and ``tearDownClass`` on a base class
```
Is it a typo? Because your overriden method is the method on a superclass,  so maybe it is something wrong that 'if you want the method on a superclass called from your superclass method'

vrpolakatcisco replied:
I found a simpler way to make it clear normal inheritance rules apply, without referring to my original use case (chain of three classes).

JulienPalard commented on Doc/library/unittest.rst:
```diff
@@ -2234,8 +2234,8 @@ Class and Module Fixtures
 the test suite encounters a test from a new class then :meth:`tearDownClass`
-from the previous class (if there is one) is called, followed by
-:meth:`setUpClass` from the new class.
+from the previous class (if it defines or inherits one) is called,
```
I don't think "defines or inheites" is clearer. What was your intent?

vrpolakatcisco replied:
> What was your intent?

I am just trying to win an argument by updating opponent's quote.
I spotted an edit [0] that I considered to me unnecessary, then started a discussion [1] on a similar edit, and my attempt to undo [2] such edits has been flatly refused.

[0] https://gerrit.fd.io/r/c/vpp/+/17092/3/test/test_bfd.py
[1] https://gerrit.fd.io/r/c/vpp/+/18579/1/test/test_sparse_vec.py#14
[2] https://gerrit.fd.io/r/c/vpp/+/18820

JulienPalard replied:
I see you're trying to win an argument, but I stil not understand how modifing this specific sentence would have helped.

JulienPalard commented on Doc/library/unittest.rst:
```diff
@@ -2269,7 +2269,7 @@ are a framework author it may be relevant.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-These must be implemented as class methods::
+These must be implemented (or inherited) as class methods::
```
Same, what are you trying to clarify? As said in b.p.o normal inheritance rules apply. I think we should be explicit about inheritance only when "anormal rule apply", not every time "normal" rules apply.

vrpolakatcisco replied:
The original formulation
  "These must be implemented as class methods"
sounds like normal rules do not apply for some reason, and each class needs to have an implementation (even if it only calls ancestor implementation).

JulienPalard replied:
> sounds like normal rules do not apply for some reason

Yes, it tells it's not a method but a class method, it reminds you not to forgot to add the `@classmethod`, the original wording is OK.

> and each class needs to have an implementation (even if it only calls ancestor implementation)

Can you proove it with an example?

JulienPalard commented on Doc/library/unittest.rst:
```diff
@@ -2282,9 +2282,8 @@ These must be implemented as class methods::
-If you want the ``setUpClass`` and ``tearDownClass`` on base classes called
-then you must call up to them yourself. The implementations in
-:class:`TestCase` are empty.
+The usual inheritance rules apply.
```
Same here, let's not be explicit about this being normal (if we're explicit here about the inheritance working normally, why not adding it everywhere else on the doc? No.)

vrpolakatcisco replied:
The original formulation
  "you must call up to them yourself"
once again sounds like normal rules do not apply for some reason, and you have to make the call explicit, instead of relying on inheritance.

> if we're explicit here about the inheritance working normally, why not adding it everywhere else on the doc?

I agree that confirming something is normal creates doubts on other (not explicitly confirmed things).

I am going to propose different edits, focusing on just removing the "wrong sounding" formulations.

JulienPalard replied:
> once again sounds like normal rules do not apply for some reason

I agree.

> and you have to make the call explicit, instead of relying on inheritance

Inheritance never "call implicitly", as long as you re-implement the method, you *have* to call the parent implementation manually. But if you don't implement a method, the parent one is just inherited and no manual actions are needed. Only normal rules here.

JulienPalard reviewed (changes requested):
Thanks for trying to enhance the documentation! I propose to simplify further, as it's only normal rules.

JulienPalard commented on Doc/library/unittest.rst:
```diff
@@ -2309,9 +2309,9 @@ Class and Module Fixtures
 
 Class and module level fixtures are implemented in :class:`TestSuite`. When
-the test suite encounters a test from a new class then :meth:`tearDownClass`
-from the previous class (if there is one) is called, followed by
```
Why removing the "if there is one"? It was right:
```
tearDownClass = getattr(previousClass, 'tearDownClass', None)
if tearDownClass is not None:
    ...
```

I don't see any problem with this paragraph, and it looks not correlated with bpo-36612.

JulienPalard commented on Doc/library/unittest.rst:
```diff
@@ -2345,7 +2345,7 @@ are a framework author it may be relevant.
 setUpClass and tearDownClass
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-These must be implemented as class methods::
```
Looks unrelated to the issue, and I don't see how the new wording helps.

JulienPalard commented on Doc/library/unittest.rst:
```diff
@@ -2372,7 +2370,7 @@ instead of as an error.
 setUpModule and tearDownModule
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-These should be implemented as functions::
```
I prefer the original wording.

vrpolakatcisco replied:
Not sure why module functions have "should" and class methods have "must", but I think I can make do edits also without affecting these two lines.

JulienPalard commented on Doc/library/unittest.rst:
```diff
@@ -2358,9 +2358,7 @@ These must be implemented as class methods::
-If you want the ``setUpClass`` and ``tearDownClass`` on base classes called
-then you must call up to them yourself. The implementations in
-:class:`TestCase` are empty.
+The implementations in :class:`TestCase` are present, but empty.
```
This sentence is right, the implementations are present but empty. But now without context I don't see the point of mentioning it, why not removing it all?

vrpolakatcisco replied:
I will expand in next version.

JelleZijlstra replied:
Honestly this still doesn't sound very useful to me. The new text doesn't really answer the question in the original bug report.

vrpolakatcisco committed:
bpo-36612: Edit unittest.rst on SetUpClass calls

The previous formulations could suggest
that some special rules apply for the class methods
as opposed to the normal inheritance rules.

The formulations are tweaked by this patch,
hinting the normal inheritance rules indeed apply.

Sentences describing module functions are also tweaked,
to signify the logic (almost) the same as for class methods.

Signed-off-by: Vratko Polak <vrpolak@cisco.com>

vrpolakatcisco commented on Doc/library/unittest.rst:
```diff
@@ -2309,17 +2309,20 @@ Class and Module Fixtures
 ``tearDownModule`` are run.
 
+The abovementioned calls are only executed,
+if the corresponding method or function is found.
```
Choosing "method is found", instead of "method exists" or "class has a method", in order to hint method resolution order is in effect.

vrpolakatcisco commented on Doc/library/unittest.rst:
```diff
@@ -2358,9 +2361,10 @@ These must be implemented as class methods::
-then you must call up to them yourself. The implementations in
-:class:`TestCase` are empty.
+The implementations in :class:`TestCase` are present, but empty.
+That means the implemenation is executed for classes descending from
```
Choosing "descending" instead of "derived" or "child", to suggest the inheritance graphs can be complicated.

vrpolakatcisco commented on Doc/library/unittest.rst:
```diff
@@ -2358,9 +2361,10 @@ These must be implemented as class methods::
-:class:`TestCase` are empty.
+The implementations in :class:`TestCase` are present, but empty.
+That means the implemenation is executed for classes descending from
+:class:`TestCase`, but the implementation does not do anything,
```
Empty implementation executed means no special handling is there to interfere with normal inheritance rules.

JelleZijlstra committed:
Merge branch 'main' into fixtures_doc

AlexWaygood committed:
Merge branch 'main' into fixtures_doc
